flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveMetaStoreFactory.java

```diff
@@ -0,0 +1,114 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.connectors.hive;
+
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.table.catalog.hive.client.HiveMetastoreClientFactory;
+import org.apache.flink.table.catalog.hive.client.HiveMetastoreClientWrapper;
+import org.apache.flink.table.catalog.hive.util.HiveTableUtil;
+import org.apache.flink.table.filesystem.MetaStoreFactory;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.thrift.TException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Optional;
+
+/**
+ * Hive {@link MetaStoreFactory}, use {@link HiveMetastoreClientWrapper} to communicate with
+ * hive meta store.
+ */
+public class HiveMetaStoreFactory implements MetaStoreFactory {
+
+	private static final long serialVersionUID = 1L;
+
+	private final JobConfWrapper conf;
+	private final String hiveVersion;
+	private final String database;
+	private final String tableName;
+
+	HiveMetaStoreFactory(
+			JobConf conf,
+			String hiveVersion,
+			String database,
+			String tableName) {
+		this.conf = new JobConfWrapper(conf);
+		this.hiveVersion = hiveVersion;
+		this.database = database;
+		this.tableName = tableName;
+	}
+
+	@Override
+	public HiveTableMetaStore createTableMetaStore() throws Exception {
+		return new HiveTableMetaStore();
+	}
+
+	private class HiveTableMetaStore implements TableMetaStore {
+
+		private HiveMetastoreClientWrapper client;
+		private StorageDescriptor sd;
+
+		private HiveTableMetaStore() throws TException {
+			client = HiveMetastoreClientFactory.create(
+					new HiveConf(conf.conf(), HiveConf.class), hiveVersion);
+			sd = client.getTable(database, tableName).getSd();
+		}
+
+		@Override
+		public Path getLocationPath() {
+			return new Path(sd.getLocation());
+		}
+
+		@Override
+		public Optional<Path> getPartition(
+				LinkedHashMap<String, String> partSpec) throws Exception {
+			try {
+				return Optional.of(new Path(client.getPartition(
+						database,
+						tableName,
+						new ArrayList<>(partSpec.values()))
+						.getSd().getLocation()));
+			} catch (NoSuchObjectException ignore) {
+				return Optional.empty();
+			}
+		}
+
+		@Override
+		public void createPartition(LinkedHashMap<String, String> partSpec, Path path) throws Exception {
+			StorageDescriptor newSd = new StorageDescriptor(sd);
+			newSd.setLocation(path.toString());
+			Partition partition = HiveTableUtil.createHivePartition(database, tableName,
+					new ArrayList<>(partSpec.values()), newSd, new HashMap<>());
+			partition.setValues(new ArrayList<>(partSpec.values()));
+			client.add_partition(partition);
```

Comment:
Don't we have to handle cases when table is not partitioned or the partition already exists?

Reply:
These two interface are just wrap `client.getPartition` and `client.add_partition`, should not exist other logical.